File: src/commands/status.ts
import { Command } from "commander";
import chalk from "chalk";
import Table from "cli-table3";
import { listBookmarks, listSpaces } from "../lib/store.js";
import { catalogPath, resolveCatalogReference } from "../lib/catalogResolver.js";
import {
  clearRuns,
  detectRunningSessions,
  findRun,
  getLatestRunForSession,
  reconcileStaleRuns,
  removeRun,
  statusBadge,
  type DetectedSession,
} from "../lib/runs.js";
import { shortSessionId } from "../lib/sessionDisplay.js";
import type { Bookmark, RunStatus } from "../types.js";

interface StatusRow {
  catalog: string;
  session_id: string;
  title: string;
  status: RunStatus;
  started_at?: string;
  ended_at?: string;
  exit_code?: number;
  pid?: number;
  source?: string;
}

function resolveStatus(
  sessionId: string,
  latest: ReturnType<typeof getLatestRunForSession>,
  detected: Map<string, DetectedSession>
): RunStatus {
  if (detected.has(sessionId)) return "running";
  if (latest && latest.status === "running") return "running"; // reconciled live pid
  return latest?.status ?? "unknown";
}

async function collectCatalogBookmarks(catalogFilter?: string): Promise<{ bookmarks: Bookmark[]; catalogError?: string }> {
  const all = listBookmarks().filter((b) => b.space_ids.length > 0);
  if (!catalogFilter) return { bookmarks: all };
  const resolution = resolveCatalogReference(catalogFilter);
  if (resolution.kind !== "found") {
    return {
      bookmarks: [],
      catalogError:
        resolution.kind === "ambiguous"
          ? `Ambiguous catalog "${catalogFilter}": ${resolution.matches.map((m) => m.name).join(", ")}`
          : `Catalog not found: ${catalogFilter}`,
    };
  }
  const catalogId = resolution.space.id;
  return { bookmarks: all.filter((b) => b.space_ids.includes(catalogId)) };
}

async function buildSnapshot(catalogFilter?: string, withDetection = false): Promise<{ rows: StatusRow[]; error?: string }> {
  reconcileStaleRuns();
  const { bookmarks, catalogError } = await collectCatalogBookmarks(catalogFilter);
  if (catalogError) return { rows: [], error: catalogError };

  const detected = withDetection ? await detectRunningSessions() : new Map<string, DetectedSession>();
  const spaces = listSpaces();
  const rows: StatusRow[] = bookmarks.map((b) => {
    const latest = getLatestRunForSession(b.session_id);
    const firstSpaceId = b.space_ids[0];
    const space = firstSpaceId ? spaces.find((s) => s.id === firstSpaceId) : undefined;
    const catalog = space ? catalogPath(space, spaces) : b.space_ids.join(",") || "-";
    return {
      catalog,
      session_id: b.session_id,
      title: b.title,
      status: resolveStatus(b.session_id, latest, detected),
      started_at: latest?.started_at,
      ended_at: latest?.ended_at,
      exit_code: latest?.exit_code,
      pid: latest?.pid,
      source: latest?.source,
    };
  });
  // Sort: running first, then by most recent started_at.
  rows.sort((a, b) => {
    if (a.status === "running" && b.status !== "running") return -1;
    if (b.status === "running" && a.status !== "running") return 1;
    return (b.started_at ?? "").localeCompare(a.started_at ?? "");
  });
  return { rows };
}

function fmtDate(value?: string): string {
  if (!value) return "-";
  return value.slice(0, 19).replace("T", " ");
}

function renderTable(rows: StatusRow[]): string {
  if (rows.length === 0) return chalk.yellow("No catalog-archived sessions.");
  const table = new Table({
    head: [
      chalk.cyan("Catalog"),
      chalk.cyan("Session"),
      chalk.cyan("Title"),
      chalk.cyan("Status"),
      chalk.cyan("Started"),
      chalk.cyan("Ended"),
      chalk.cyan("Exit"),
      chalk.cyan("PID"),
    ],
    colWidths: [18, 14, 24, 14, 20, 20, 7, 9],
    style: { head: [] },
  });
  for (const r of rows) {
    const title = r.title.length > 22 ? r.title.slice(0, 22) + "…" : r.title;
    table.push([
      r.catalog.length > 16 ? "…" + r.catalog.slice(-15) : r.catalog,
      shortSessionId(r.session_id),
      title,
      `${statusBadge(r.status)} ${r.status}`,
      fmtDate(r.started_at),
      fmtDate(r.ended_at),
      r.exit_code === undefined ? "-" : String(r.exit_code),
      r.pid === undefined ? "-" : String(r.pid),
    ]);
  }
  return table.toString();
}

function clearScreen(): void {
  process.stdout.write("\x1Bc");
}

export function registerStatusCommand(program: Command): void {
  const status = new Command("status")
    .description("Monitor the run status of catalog-archived sessions")
    .argument("[catalog]", "filter to a catalog (name, path, or id)")
    .option("-c, --catalog <catalog>", "filter to a catalog (name, path, or id)")
    .option("--watch", "live monitoring mode (re-render every few seconds)")
    .option("--json", "output current snapshot as JSON")
    .action(async (arg: string | undefined, opts: { catalog?: string; watch?: boolean; json?: boolean }) => {
      const catalogFilter = opts.catalog ?? arg;

      if (opts.json) {
        const { rows, error } = await buildSnapshot(catalogFilter, true);
        if (error) {
          console.error(chalk.red(error));
          process.exit(1);
        }
        console.log(JSON.stringify(rows, null, 2));
        return;
      }

      if (!opts.watch) {
        const { rows, error } = await buildSnapshot(catalogFilter, true);
        if (error) {
          console.error(chalk.red(error));
          process.exit(1);
        }
        console.log(renderTable(rows));
        return;
      }

      // Watch mode.
      let previous = new Map<string, RunStatus>();
      let tick = 0;
      let stopped = false;
      const stop = () => {
        if (stopped) return;
        stopped = true;
        clearScreen();
        process.exit(0);
      };
      process.on("SIGINT", stop);

      const renderOnce = async () => {
        tick++;
        const withDetection = tick % 2 === 1; // detect every other tick (~6s)
        const { rows } = await buildSnapshot(catalogFilter, withDetection);
        const current = new Map<string, RunStatus>(rows.map((r) => [r.session_id, r.status]));

        const events: string[] = [];
        for (const [sid, status] of current) {
          const prev = previous.get(sid);
          if (prev === undefined) continue;
          if (prev !== status) {
            events.push(`[${new Date().toISOString().slice(11, 19)}] session ${shortSessionId(sid)} ${prev} → ${status}`);
          }
        }
        previous = current;

        clearScreen();
        const filterLine = catalogFilter ? chalk.gray(`catalog: ${catalogFilter}\n`) : "";
        process.stdout.write(filterLine + renderTable(rows) + "\n");
        if (events.length > 0) {
          process.stdout.write(chalk.gray("\n— transitions —\n") + events.map((e) => chalk.gray(e)).join("\n") + "\n");
        }
        process.stdout.write(chalk.gray(`\nrefreshing… (Ctrl-C to exit, tick ${tick})\n`));
      };

      await renderOnce();
      const interval = setInterval(() => {
        renderOnce().catch(() => undefined);
      }, 3000);
      interval.unref?.();
    });

  status
    .command("prune")
    .description("Remove crashed and stale run records")
    .action(() => {
      const crashed = clearRuns({ status: "crashed" });
      const stale = clearRuns({ status: "stale" });
      console.log(chalk.green(`Pruned ${crashed + stale} run record(s) (crashed: ${crashed}, stale: ${stale}).`));
    });

  status
    .command("clear <run-or-session>")
    .description("Remove a run record by run_id or all records for a session_id")
    .action((target: string) => {
      const run = findRun(target);
      if (run) {
        removeRun(target);
        console.log(chalk.green(`Removed run ${target}.`));
        return;
      }
      const removed = clearRuns({ session_id: target });
      if (removed > 0) {
        console.log(chalk.green(`Removed ${removed} run record(s) for session ${target}.`));
      } else {
        console.log(chalk.yellow(`No run records matched "${target}".`));
      }
    });

  program.addCommand(status);
}

File: rust/src/commands/status.rs
//! `starling status` — show pinned sessions and their run status.

use std::collections::HashMap;

use anyhow::Result;
use colored::*;

use crate::cli::*;
use crate::core::catalog_resolver::{catalog_path, resolve_catalog_reference, CatalogResolution};
use crate::core::runs::{
    clear_runs, detect_running_sessions, get_latest_run_for_session, reconcile_stale_runs, status_badge,
    ListFilter, RunFilter, RunStatus,
};
use crate::core::session_display::short_session_id;
use crate::core::store::{list_bookmarks, list_spaces, BookmarkFilter};
use crate::core::process_metrics::get_process_tree_metrics;
use crate::types::Bookmark;

#[derive(Debug, serde::Serialize)]
struct StatusRow {
    catalog: String,
    session_id: String,
    title: String,
    status: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    started_at: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    ended_at: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    exit_code: Option<i32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pid: Option<u32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    source: Option<String>,
}

pub fn handle(cmd: StatusCommand) -> Result<()> {
    match cmd {
        StatusCommand::Show { catalog, live, json } => show(catalog, live, json),
        StatusCommand::Prune { json } => prune(json),
        StatusCommand::Clear { yes, json } => clear_cmd(yes, json),
    }
}

fn collect_catalog_bookmarks(catalog_filter: Option<&str>) -> (Vec<Bookmark>, Option<String>) {
    let mut all: Vec<Bookmark> = list_bookmarks(BookmarkFilter::default()).into_iter()
        .filter(|b| !b.space_ids.is_empty())
        .collect();
    let Some(filter) = catalog_filter else { return (all, None); };
    match resolve_catalog_reference(filter) {
        CatalogResolution::Found(space) => {
            all.retain(|b| b.space_ids.contains(&space.id));
            (all, None)
        }
        CatalogResolution::Ambiguous(matches) => {
            (vec![], Some(format!("Ambiguous catalog \"{}\": {}", filter,
                matches.iter().map(|s| s.name.clone()).collect::<Vec<_>>().join(", "))))
        }
        CatalogResolution::NotFound => (vec![], Some(format!("Catalog not found: {}", filter))),
    }
}

fn build_snapshot(catalog_filter: Option<&str>, with_detection: bool) -> (Vec<StatusRow>, Option<String>) {
    reconcile_stale_runs();
    let (bookmarks, error) = collect_catalog_bookmarks(catalog_filter);
    if let Some(err) = error { return (vec![], Some(err)); }

    let detected = if with_detection {
        detect_running_sessions()
    } else {
        HashMap::new()
    };
    let spaces = list_spaces();
    let rows: Vec<StatusRow> = bookmarks.iter().map(|b| {
        let latest = get_latest_run_for_session(&b.session_id);
        let first_space_id = b.space_ids.first();
        let space = first_space_id.and_then(|sid| spaces.iter().find(|s| &s.id == sid));
        let catalog = space.map(|s| catalog_path(s, Some(&spaces)))
            .unwrap_or_else(|| b.space_ids.join(","));
        let status = if detected.contains_key(&b.session_id) {
            "running".to_string()
        } else if latest.as_ref().map(|r| r.status == RunStatus::Running).unwrap_or(false) {
            "running".to_string()
        } else {
            latest.as_ref().map(|r| format!("{:?}", r.status).to_lowercase()).unwrap_or_else(|| "unknown".to_string())
        };
        StatusRow {
            catalog,
            session_id: b.session_id.clone(),
            title: b.title.clone(),
            status,
            started_at: latest.as_ref().map(|r| r.started_at.clone()),
            ended_at: latest.as_ref().and_then(|r| r.ended_at.clone()),
            exit_code: latest.as_ref().and_then(|r| r.exit_code),
            pid: latest.as_ref().and_then(|r| r.pid),
            source: latest.as_ref().map(|r| format!("{:?}", r.source).to_lowercase()),
        }
    }).collect();
    (rows, None)
}

fn show(catalog_filter: Option<String>, live: bool, json: bool) -> Result<()> {
    let (rows, error) = build_snapshot(catalog_filter.as_deref(), live);
    if let Some(err) = error {
        eprintln!("{}: {}", "error".red(), err);
        std::process::exit(2);
    }
    if json {
        println!("{}", serde_json::to_string_pretty(&rows)?);
        return Ok(());
    }
    if rows.is_empty() {
        println!("{}", "No cataloged sessions found.".yellow());
        return Ok(());
    }
    use comfy_table::{Cell, Color, ContentArrangement, Table, presets::UTF8_FULL};
    let mut table = Table::new();
    table.load_preset(UTF8_FULL).set_content_arrangement(ContentArrangement::Disabled);
    table.set_header(vec![
        Cell::new("ST").fg(Color::Cyan),
        Cell::new("Session").fg(Color::Cyan),
        Cell::new("Catalog").fg(Color::Cyan),
        Cell::new("Title").fg(Color::Cyan),
        Cell::new("Started").fg(Color::Cyan),
        Cell::new("CPU%").fg(Color::Cyan),
        Cell::new("RSS(MB)").fg(Color::Cyan),
        Cell::new("PID").fg(Color::Cyan),
    ]);
    for r in &rows {
        let st = parse_status(&r.status);
        let mut cpu = String::new();
        let mut rss_mb = String::new();
        let mut pid_str = String::new();
        if let Some(pid) = r.pid {
            pid_str = pid.to_string();
            let m = get_process_tree_metrics(pid);
            cpu = format!("{:.1}", m.cpu_pct);
            rss_mb = format!("{:.0}", m.mem_kb as f64 / 1024.0);
        }
        table.add_row(vec![
            Cell::new(status_badge(st, false)),
            Cell::new(short_session_id(&r.session_id)),
            Cell::new(&r.catalog),
            Cell::new(&r.title),
            Cell::new(r.started_at.as_deref().unwrap_or("-")),
            Cell::new(cpu),
            Cell::new(rss_mb),
            Cell::new(pid_str),
        ]);
    }
    println!("{table}");
    Ok(())
}

fn parse_status(s: &str) -> RunStatus {
    match s {
        "running" => RunStatus::Running,
        "completed" => RunStatus::Completed,
        "errored" => RunStatus::Errored,
        "crashed" => RunStatus::Crashed,
        "stale" => RunStatus::Stale,
        _ => RunStatus::Unknown,
    }
}

fn prune(json: bool) -> Result<()> {
    let n = reconcile_stale_runs();
    if json {
        return super::print_json_result(
            "status.prune",
            &format!("Reconciled {} stale run(s).", n),
            serde_json::json!({ "reconciled": n }),
        );
    }
    println!("{}", format!("Reconciled {} stale run(s).", n).green());
    Ok(())
}

fn clear_cmd(yes: bool, json: bool) -> Result<()> {
    if !yes {
        eprintln!("{}: clearing all run records requires --yes", "error".red());
        std::process::exit(2);
    }
    let removed = clear_runs(None);
    let _ = ListFilter::default();
    let _ = RunFilter::default();
    if json {
        return super::print_json_result(
            "status.clear",
            &format!("Cleared {} run record(s).", removed),
            serde_json::json!({ "removed": removed }),
        );
    }
    println!("{}", format!("Cleared {} run record(s).", removed).green());
    Ok(())
}
